File: go/mqtt/retrypolicy/exponential_backoff_retrypolicy.go
package retrypolicy

import (
	"context"
	"fmt"
	"math"
	"math/rand"
	"time"

	"github.com/Azure/iot-operations-sdks/go/internal/wallclock"
)

const (
	// Start with an exponent of 1 for the delay calculation,
	// which begins too low and takes too long to exceed 1 second.
	// Adding 6 to the retry count starts at 2^7 = 128 milliseconds,
	// exceeding 1 second on the 4th retry.
	minExponent = 6

	// Avoid integer overflow by clamping the maximum delay.
	// The maximum exponent value is set to 32.
	maxExponent = 32
)

// ExponentialBackoffRetry implements a retry policy
// with exponential backoff and optional jitter.
type ExponentialBackoffRetryPolicy struct {
	// If maxRetries is nil, we start indefinitely retry.
	maxRetries  *int
	maxInterval time.Duration
	// If timeout is 0, we start indefinitely retry.
	timeout    time.Duration
	withJitter bool
}

// NewExponentialBackoffRetryPolicy creates
// a new ExponentialBackoffRetryPolicy with the given options.
func NewExponentialBackoffRetryPolicy(
	opts ...ExponentialBackoffRetryPolicyOption,
) *ExponentialBackoffRetryPolicy {
	exponentialBackoffRetryPolicy := &ExponentialBackoffRetryPolicy{
		maxRetries:  nil, // Set indefinitely retry by default
		maxInterval: defaultMaxInterval,
		timeout:     0, // Set indefinitely retry by default
		withJitter:  defaultWithJitter,
	}

	for _, opt := range opts {
		opt(exponentialBackoffRetryPolicy)
	}

	return exponentialBackoffRetryPolicy
}

// Start initiates the retry executions.
func (e *ExponentialBackoffRetryPolicy) Start(
	ctx context.Context,
	log func(msg string, args ...any),
	task Task,
) error {
	var retryCtx context.Context
	var cancel context.CancelFunc

	// Create a context with timeout if specified
	if e.timeout != 0 {
		retryCtx, cancel = context.WithTimeout(ctx, e.timeout)
		defer cancel()
	} else {
		retryCtx = ctx
	}

	for try := 0; ; try++ {
		log(fmt.Sprintf(
			"retry: executing %s on attempt %d",
			task.Name,
			try+1,
		))
		err := task.Exec(retryCtx)
		if err == nil {
			e.status(log, task.Name, try, nil)
			return nil
		}

		interval := e.shouldRetry(retryCtx, try, task.Cond(err))
		if interval == 0 {
			e.status(log, task.Name, try, err)
			return err
		}

		select {
		case <-wallclock.Instance.After(interval):
		case <-retryCtx.Done():
			e.status(log, task.Name, try, retryCtx.Err())
			return retryCtx.Err()
		}
	}
}

// shouldRetry decides if we need to continue/start
// retrying the target operations
// based on retry count and other conditions.
func (e *ExponentialBackoffRetryPolicy) shouldRetry(
	ctx context.Context,
	retries int,
	cond bool,
) time.Duration {
	if e.maxRetries != nil &&
		(*e.maxRetries <= 0 ||
			retries >= *e.maxRetries-1 ||
			!cond ||
			ctx.Err() != nil) {
		return 0
	}

	// Calculate exponent and clamp to max exponent
	exp := retries + minExponent
	if exp > maxExponent {
		exp = maxExponent
	}
	expIntervalMs := math.Pow(2.0, float64(exp))
	clampedMs := math.Min(expIntervalMs, float64(e.maxInterval.Milliseconds()))
	interval := time.Duration(clampedMs) * time.Millisecond

	if e.withJitter {
		interval = e.jitter(clampedMs)
	}

	return interval
}

// status returns the result of each retry.
func (*ExponentialBackoffRetryPolicy) status(
	log func(msg string, args ...any),
	task string,
	try int,
	err error,
) {
	if err != nil {
		log(fmt.Sprintf(
			"retry: %s failed after %d attempt(s): %v",
			task,
			try+1,
			err,
		))
	} else {
		log(fmt.Sprintf(
			"retry: %s succeeded after %d attempt(s)",
			task,
			try+1,
		))
	}
}

// jitter adds a random jitter to the base time to avoid synchronicity
// in retry attempts.
// The jitter is between 95% and 105% of the base time.
func (*ExponentialBackoffRetryPolicy) jitter(base float64) time.Duration {
	// Generate a random jitter value between 0 and 10
	seed := wallclock.Instance.Now().UnixNano()
	// #nosec G404
	r := rand.New(rand.NewSource(seed))
	j := r.Intn(11)

	// Calculate jitter percentage by adding 95 to the random value,
	// resulting in a range of 95 to 105
	percent := j + 95

	// Apply the jitter percentage to the base time
	jitter := base * float64(percent) / 100.0

	return time.Duration(jitter) * time.Millisecond
}

File: dotnet/src/Azure.Iot.Operations.Protocol/Retry/ExponentialBackoffRetryPolicy.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System;

namespace Azure.Iot.Operations.Protocol.Retry
{
    /// <summary>
    /// Implements Binary Exponential Backoff (BEB) retry policy as follows:
    /// CurrentExponent = min(MaxExponent, baseExponent + currentRetryCount)
    /// RetryDelay = min(pow(2, CurrentExponent), maxWait) milliseconds
    /// </summary>
    public class ExponentialBackoffRetryPolicy : IRetryPolicy
    {
        // Default base exponent set equal 6, in this case first retry starts at 2^(6+1)=128 milliseconds, and exceed 1 second delay on retry #4.
        private readonly uint _baseExponent = 6u;

        // Avoid integer overflow (max of 32) and clamp max delay.
        private const uint MaxExponent = 32u;

        /// <summary>
        /// The maximum number of retries
        /// </summary>
        private readonly uint _maxRetries;

        private readonly TimeSpan _maxDelay;
        private readonly bool _useJitter;

        /// <summary>
        /// Creates an instance of this class with a default base exponent equals 6.
        /// </summary>
        /// <param name="maxRetries">The maximum number of retry attempts.</param>
        /// <param name="maxWait">The maximum amount of time to wait between retries.</param>
        /// <param name="useJitter">Whether to add a small, random adjustment to the retry delay to avoid synchronicity in clients retrying.</param>
        public ExponentialBackoffRetryPolicy(uint maxRetries, TimeSpan maxWait, bool useJitter = true)
        {
            _maxRetries = maxRetries;
            _maxDelay = maxWait;
            _useJitter = useJitter;
        }

        /// <summary>
        /// Creates an instance of this class.
        /// </summary>
        /// <param name="maxRetries">The maximum number of retry attempts.</param>
        /// <param name="baseExponent">The base exponent to start the backoff calculation (CurrentExponent(currentRetryCount) = baseExponent + currentRetryCount).</param>
        /// <param name="maxWait">The maximum amount of time to wait between retries.</param>
        /// <param name="useJitter">Whether to add a small, random adjustment to the retry delay to avoid synchronicity in clients retrying.</param>
        public ExponentialBackoffRetryPolicy(uint maxRetries, uint baseExponent, TimeSpan maxWait, bool useJitter = true) :
        this(maxRetries, maxWait, useJitter)
        {
            _baseExponent = baseExponent;
        }

        /// <inheritdoc/>
        public bool ShouldRetry(uint currentRetryCount, Exception? lastException, out TimeSpan retryDelay)
        {
            if (_maxRetries == 0 || currentRetryCount > _maxRetries)
            {
                retryDelay = TimeSpan.Zero;
                return false;
            }

            // Avoid integer overflow and clamp max delay.
            // if currentRetryCount is very high, adding MinExponent would just wrap around safely
            // and decrease the value suddenly, so exponent gets capped before addition
            // Result: The delay stays at maximum instead of suddenly dropping.
            uint exponent;
            if (currentRetryCount > uint.MaxValue - _baseExponent)
            {
                exponent = MaxExponent;
            }
            else
            {
                exponent = currentRetryCount + _baseExponent;
                exponent = Math.Min(MaxExponent, exponent);
            }

            // 2 to the power of the retry count gives us exponential back-off.
            double exponentialIntervalMs = Math.Pow(2.0, exponent);

            double clampedWaitMs = Math.Min(exponentialIntervalMs, _maxDelay.TotalMilliseconds);

            retryDelay = _useJitter
                ? UpdateWithJitter(clampedWaitMs)
                : TimeSpan.FromMilliseconds(clampedWaitMs);

            return true;
        }

        /// <summary>
        /// Gets jitter between 95% and 105% of the base time.
        /// </summary>
        private TimeSpan UpdateWithJitter(double baseTimeMs)
        {
            // Don't calculate jitter if the value is very small
            if (baseTimeMs < 50)
            {
                return TimeSpan.FromMilliseconds(baseTimeMs);
            }

            double jitterMs = Random.Shared.Next(95, 106) * baseTimeMs / 100.0;

            return TimeSpan.FromMilliseconds(jitterMs);
        }
    }
}
